cmd/containerd-shim-runhcs-v1/serve.go

```diff
@@ -232,12 +232,25 @@ var serveCommand = cli.Command{
 
 			// This is our best indication that we have not errored on creation
 			// and are successfully serving the API.
+			// Closing stdout signals to containerd that shim started successfully
 			os.Stdout.Close()
 		}
 
 		// Wait for the serve API to be shut down.
-		<-serrs
-		return nil
+		select {
+		case err = <-serrs:
+			// the ttrpc server shutdown without processing a shutdown request
+		case <-svc.Done():
+			if !svc.GracefulShutdown {
+				// Return immediately, but still close ttrpc server, pipes, and spans
+				// Shouldn't need to os.Exit without clean up (ie, deferred `.Close()`s)
+				return nil
+			}
+			// Drain any remaining active ttrpc requests; times out after 200 ms
+			err = s.Shutdown(context.Background())
```

Comment:
This assumption is incorrect. The ticker is at 200ms intervals for ttrpc.Shutdown which just controls the loop variants assuming all connections get closed properly. What you really need to do is set a context timeout here on the passed in context that will forcibly exit the loop after whatever time you want. So while the loop makes progress and is less than your time it will continue to gracefully shutdown connections. But after that will forcibly kill the shim

Reply:
I misreach `ticker` as `timer`, which is ... embarrassing.
Updated to use a `context.WithTimeout` to timebox the shutdown

Reply:
No prob, easy to do.